tracing-journald/src/socket.rs

```diff
@@ -0,0 +1,65 @@
+//! socket helpers.
+
+use std::alloc::{alloc_zeroed, dealloc, Layout};
+use std::io::{Error, Result};
+use std::mem::{align_of, size_of, MaybeUninit};
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::prelude::{AsRawFd, RawFd};
+use std::ptr;
+
+use libc::*;
+
+pub fn send_one_fd<F: AsRawFd>(socket: &UnixDatagram, fd: F) -> Result<usize> {
+    unsafe {
+        // We don't initialize msghdr because the musl 64bit implementation contains private padding fields.
+        // We zero this chunk of memory to make sure that all fields we do not initialize explicitly and all
+        // padding fields are zero.
+        let mut msg: msghdr = MaybeUninit::zeroed().assume_init();
+
+        // Allocate the buffer for the control part of the message on the heap with proper alignment.
+        // Ideally we'd use a stack-allocated repr(C) union for the buffer and proper alignment,
+        // as outlined in cmsg(3), but `CMSG_SPACE` is no const function, unfortunately (the corresponding
+        // libc feature `const-extern-fn` requires nightly).
+        //
+        // We allocate zeroed memory to make sure that all padding bytes in CMSG_SPACE are zero.
+        let control_layout = Layout::from_size_align(
+            CMSG_SPACE(size_of::<RawFd>() as u32) as usize,
+            align_of::<cmsghdr>(),
+        )
+        .unwrap();
+        let controlbuf = alloc_zeroed(control_layout);
```

Comment:
We should put this on the stack. It's okay to hardcode an empirical value based on the behavior of `CMSG_SPACE` on known platforms. We can check the assumption in `Subscriber::new` for safety.

Reply:
So you'd like me to go… through all combinations of libc's and architectures to hardcode the values of CMSG_SPACE for these?

I don't think I'd like to do this, to be honest.  I'm sorry but that more effort than I'd like to spend on this merge request, and I don't see the benefit: While I do agree that a stack allocation would be nice and easier to understand this code only gets hit in the rare situation where someone actually sends a few megabytes to the journal.  I don't think this corner case warrants so much effort, and I figure that if you send a couple of megabytes to the journal this little heap allocation won't do no further harm :slightly_smiling_face: 

If you think that it's quite important to do a heap allocation here let's at least do it "right" and try to get libc to promote CMSG_SPACE to a const fn; as far as I can see it's not an extern function so there's no reason why it shouldn't be const  by default.

Reply:
FWIW nix also uses a heap-allocated `Vec`, for the same reason.  

I tend to put the blame on `libc`, to be honest: There's no reason for `CMSG_SPACE` and `CMSG_ALIGN` to be guarded under the const extern fn feature; they aren't extern at all and could trivially be propagated to `const`.

Reply:
You're right that the performance on this path isn't super important. However, we don't need to determine and hardcode a precise unique size for every possible target, just come up with a plausible upper bound and be prepared to raise it in the future if someone complains. `CMSG_SPACE` is de facto very straightforward, after all--it's just the sum of sizeof(cmsghdr), aligned, with the size of the payload, aligned. If you don't want to think about it any harder, I'm sure a 64 byte buffer will work just fine; we use 88 bytes in Quinn for up to 3 different cmsgs.

Reply:
Oh, I must admit the thought of just not bothering with `CMSG_SPACE` and using a fixed-size buffer instead just did not at all occur to me :innocent: I'll try that right away :slightly_smiling_face:

Reply:
I have a slight preference for doing this the way Ralith suggests, not because of the performance impact of the heap allocation, but because I think this code would be a bit simpler if we could just use a `repr(C)` union here. But, I agree that it would also be quite unpleasant to hardcode every possible `CMSG_SPACE` value here. I think it's probably fine, though, to just pick the maximum size and alignment that any platform needs, and go with that --- if we waste a few bytes of stack space on other platforms, I don't think that's the end of the world.

Reply:
@Ralith @hawkw I do agree, and I like the idea of using a fixed-size buffer.

Now, I've tried to implement it but it didn't quite work, and my C & systems foo isn't good enough to figure out what's going on :thinking: 

I'm resolving this conversation, and have added some details at the appropriate place.